straxen/plugins/acqmon_processing.py

```diff
@@ -240,6 +242,160 @@ def compute(self, aqmon_hits, start, end):
+    """
+        
+    __version__ ='0.1.0'
+    depends_on = ('aqmon_hits', 'event_info')
+    provides  = 'gps_sync'
+    data_kind = 'gps_sync'
+    
+    def infer_dtype(self):
+        dtype = [(('Time since unix epoch [ns]', 'time'), np.int64),
+                 (('GPS absolute time [ns]', 't_gps'), np.int64),
+                ]
+        return dtype
+
+
+    def gps_times_from_runid(self,run_id):
+        '''
+        Fetches the mongodb looking for the gps_sync collection for 
+        timestamps between the start and end times of a given run_id.
+        '''
+
+        rundb = utilix.xent_collection()
+        gps_times = utilix.xent_collection(collection='gps_sync',  database='xenonnt')
+
+        if isinstance(run_id, str):
+            run_id = int(run_id)
+
+        query = {"number": run_id}
+
+        doc = rundb.find_one(query, projection={'start': 1, 'end': 1})
+        
+        start_t = doc['start'].replace(tzinfo=datetime.timezone.utc).timestamp()
+        end_t = doc['end'].replace(tzinfo=datetime.timezone.utc).timestamp()
+        
+        query = {"gps_sec": { '$gte': start_t - 1, '$lte': end_t + 1}, 'channel': 0 }
+
+        return pd.DataFrame(gps_times.find(query))
+
+    def cut_outside_run(self,aqmon_array, gps_array):
+        first_idx = None
+        last_idx = None
+        _idx = 0
+        while (first_idx == None) or (last_idx == None):
+
+            if np.abs(aqmon_array[0]-gps_array[_idx])<5e9:
+                first_idx = _idx
+            if np.abs(aqmon_array[-1]-gps_array[-_idx-1])<5e9:
+                last_idx = -_idx
+            _idx+=1
+        if last_idx == 0:
+            gps_array_corr = gps_array[first_idx:]
+        else:
+            gps_array_corr = gps_array[first_idx:last_idx]
+
+        return gps_array_corr
+
+    # Basic algorithm to compute corrected time given all the inputs
+    def compute_time_individual(self, t0_daq, t1_daq, t0_gps, t1_gps, t_evt):
+        '''
+        Compute the corrected time given the previous and next sync pulse.
+          * t0: previous pulse (daq or gps)
+          * t1: previous pulse (daq or gps)
+          * t_evt: uncorrected time of event (in DAQ time)
+        Returns a single value, t_evt in gps-time.
+        '''
+        delta_sync_offset = t0_gps-t0_daq
+        delta_t_evt = t_evt - t0_daq
+        
+        return t0_daq + delta_sync_offset + (t1_gps-t0_gps)/(t1_daq-t0_daq)*delta_t_evt
+
+    def compute_time_array(self, l_daq_sync, l_gps_sync, l_daq_evt):
+        l_gps_evt = np.zeros(len(l_daq_evt))
+        
+        #correlate the values of t_daq_evt with the array l_daq_sync
+        l_idx_evt_sync_daq = np.searchsorted(l_daq_sync, l_daq_evt)
+
+        for counter,_t_evt in enumerate(l_daq_evt):
+            index_in_matched_pulses = l_idx_evt_sync_daq[counter]
+            
+            if index_in_matched_pulses == 0: # events before the first pulse
+                _t0_daq = l_daq_sync[0] - 10*1e9
```

Comment:
Are we sure about this exact 10s offset? We know the DAQ clock drifts at 0.2ppm, we should be off by 2us in this interval. (Same for L341)

Reply:
That's right, we do know it's not exactly 10s. This is an approximation to correct the events happening before the first pulse (and the analogous after the last). I could change to use the two next ones and extrapolate and that might be a more reliable and correct way.